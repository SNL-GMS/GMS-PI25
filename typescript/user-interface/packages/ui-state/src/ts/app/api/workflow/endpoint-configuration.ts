import { Endpoints } from '@gms/common-model';
import { UI_URL } from '@gms/common-util';
import type { RequestConfig, ServiceDefinition } from '@gms/ui-workers';

import { prioritizeRequests } from '../request-priority';

/**
 * Workflow request config definition
 */
export interface WorkflowRequestConfig extends RequestConfig {
  readonly workflow: {
    readonly baseUrl: string;
    readonly services: {
      readonly workflow: ServiceDefinition;
      readonly stageIntervalsByIdAndTime: ServiceDefinition;
      readonly updateActivityIntervalStatus: ServiceDefinition;
      readonly updateStageIntervalStatus: ServiceDefinition;
    };
  };
}

/**
 * The workflow request config for all services.
 */
export const config: WorkflowRequestConfig = {
  workflow: {
    baseUrl: `${UI_URL}${Endpoints.WorkflowManagerServiceUrls.baseUrl}`,
    services: prioritizeRequests({
      workflow: {
        friendlyName: Endpoints.WorkflowManagerServiceUrls.workflow.friendlyName,
        requestConfig: {
          method: 'post',
          url: Endpoints.WorkflowManagerServiceUrls.workflow.url,
          responseType: 'json',
          proxy: false,
          headers: {
            accept: 'application/json',
            'content-type': 'text/plain'
          },
          timeout: 180000,
          data: `"PlaceHolder"`
        }
      },
      stageIntervalsByIdAndTime: {
        friendlyName: Endpoints.WorkflowManagerServiceUrls.stageIntervalsByIdAndTime.friendlyName,
        requestConfig: {
          method: 'post',
          url: Endpoints.WorkflowManagerServiceUrls.stageIntervalsByIdAndTime.url,
          responseType: 'json',
          proxy: false,
          headers: {
            accept: 'application/json',
            'content-type': 'application/json'
          },
          timeout: 180000
        }
      },
      updateActivityIntervalStatus: {
        friendlyName:
          Endpoints.WorkflowManagerServiceUrls.updateActivityIntervalStatus.friendlyName,
        requestConfig: {
          method: 'post',
          url: Endpoints.WorkflowManagerServiceUrls.updateActivityIntervalStatus.url,
          responseType: 'json',
          proxy: false,
          headers: {
            accept: 'application/json',
            'content-type': 'application/json'
          },
          timeout: 60000
        }
      },
      updateStageIntervalStatus: {
        friendlyName: Endpoints.WorkflowManagerServiceUrls.updateStageIntervalStatus.friendlyName,
        requestConfig: {
          method: 'post',
          url: Endpoints.WorkflowManagerServiceUrls.updateStageIntervalStatus.url,
          responseType: 'json',
          proxy: false,
          headers: {
            'content-type': 'application/json'
          },
          timeout: 60000
        }
      }
    })
  }
};

export type WorkflowServices = keyof typeof config.workflow.services;
