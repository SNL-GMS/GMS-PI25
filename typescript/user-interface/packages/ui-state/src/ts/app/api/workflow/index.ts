export * from './workflow-api-slice';
