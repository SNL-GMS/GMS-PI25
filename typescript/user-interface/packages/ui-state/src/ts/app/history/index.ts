export * from './actions';
export * from './history-hooks';
export * from './history-keyboard-hooks';
export * from './history-selectors';
export * from './history-slice';
export * from './middleware';
export * from './reducers';
