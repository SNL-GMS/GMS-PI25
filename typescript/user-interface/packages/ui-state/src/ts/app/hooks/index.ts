export * from './action-target-hooks';
export * from './beamforming-template-hooks';
export * from './channel-hooks';
export * from './channel-segment-hooks';
export * from './data-acquisition-hooks';
export * from './event-hooks';
export * from './event-manager-hooks';
export * from './filter-hooks';
export * from './fk-channel-segment-hooks';
export * from './keyboard-shortcuts-hooks';
export * from './operational-time-period-configuration-hooks';
export * from './processing-analyst-configuration-hooks';
export * from './processing-mask-hooks';
export * from './processing-station-hooks';
export * from './qc-segment-hooks';
export * from './query-util-hooks';
export * from './react-redux-hooks';
export * from './signal-detection-hooks';
export * from './signal-enhancement-configuration-hooks';
export * from './station-definition-hooks';
export * from './system-event-gateway-hooks';
export * from './update-signal-detection-hooks';
export * from './user-manager-hooks';
export * from './user-session-hooks';
export * from './waveform-hooks';
export * from './workflow-hooks';
