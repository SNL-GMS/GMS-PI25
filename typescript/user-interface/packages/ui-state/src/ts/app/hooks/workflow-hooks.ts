import { WorkflowTypes } from '@gms/common-model';
import React from 'react';

import {
  useStageIntervalsByIdAndTimeQuery,
  useUpdateActivityIntervalStatusMutation,
  useUpdateStageIntervalStatusMutation,
  useWorkflowQuery
} from '../api';
import {
  closeStage,
  getStageName,
  openOrCloseStage,
  selectOpenActivityNames,
  selectOpenIntervalName,
  selectUsername,
  selectWorkflowTimeRange,
  updateStage
} from '../state';
import { useOperationalTimePeriodConfiguration } from './operational-time-period-configuration-hooks';
import { useAppDispatch, useAppSelector } from './react-redux-hooks';

/**
 * @returns the result of the useStageIntervalsByIdAndTimeQuery query with the stages from the workflow query,
 * and the time period from the operational time period configuration from processing config
 */
export function useStageIntervalsQuery() {
  const workflowQuery = useWorkflowQuery();

  const stageNames = React.useMemo(
    () =>
      workflowQuery.isSuccess ? workflowQuery.data?.stages?.map(stage => stage.name) ?? [] : [],
    [workflowQuery.isSuccess, workflowQuery.data?.stages]
  );

  const { timeRange } = useOperationalTimePeriodConfiguration();
  return useStageIntervalsByIdAndTimeQuery(stageNames, timeRange);
}

/**
 * Gets the stage id for the currently open interval.
 * Returns undefined if no stage interval is found, or if no interval is open.
 */
export function useStageId(): WorkflowTypes.IntervalId | undefined {
  const { data: workflowIntervals } = useStageIntervalsQuery() ?? {};
  const openIntervalName = useAppSelector(selectOpenIntervalName);
  const openTimeRange = useAppSelector(selectWorkflowTimeRange);
  if (openIntervalName == null || openTimeRange?.startTimeSecs == null) {
    return undefined;
  }
  const stageId = workflowIntervals
    ?.filter(interval => {
      return interval.name === openIntervalName;
    })
    .flatMap(stageInt => stageInt.value)
    .find(
      stageInt =>
        stageInt.intervalId.definitionId.name === openIntervalName &&
        stageInt.intervalId.startTime === openTimeRange.startTimeSecs &&
        stageInt.endTime === openTimeRange.endTimeSecs
    )?.intervalId;

  if (stageId == null) {
    // Handle the Open Anything case by building a stage interval ID
    return {
      startTime: openTimeRange.startTimeSecs,
      definitionId: {
        name: openIntervalName
      }
    };
  }
  return stageId;
}

export function useSetInterval(
  status: WorkflowTypes.IntervalStatus.NOT_COMPLETE | WorkflowTypes.IntervalStatus.IN_PROGRESS
) {
  const dispatch = useAppDispatch();
  const workflowQuery = useWorkflowQuery();
  const userName = useAppSelector(selectUsername);
  const openIntervalName = useAppSelector(selectOpenIntervalName);
  const openActivityNames = useAppSelector(selectOpenActivityNames);
  const workflowTimeRange = useAppSelector(selectWorkflowTimeRange);
  const [activityMutation] = useUpdateActivityIntervalStatusMutation();
  const [analystStageMutation] = useUpdateStageIntervalStatusMutation();

  const { data: workflow } = workflowQuery;
  const workflowPanel = document.getElementsByClassName('workflow-panel');
  return React.useCallback(
    async (interval: WorkflowTypes.ActivityInterval | WorkflowTypes.StageInterval) => {
      const { startTimeSecs } = workflowTimeRange;
      const stageName = getStageName(interval);
      const stage: WorkflowTypes.Stage = workflow?.stages.find(s => s.name === stageName);

      // Operation 1: Determines if there is activity for the currently-open stage
      const isActivityInOpenStageInterval = !(
        status === WorkflowTypes.IntervalStatus.IN_PROGRESS &&
        openIntervalName &&
        (openIntervalName !== stageName || interval.intervalId.startTime !== startTimeSecs)
      );

      // Close the stage, if applicable
      if (!isActivityInOpenStageInterval) {
        // Close the open stage and/or activities since user is discarding changes and opening another interval
        await closeStage(userName, startTimeSecs, openIntervalName, analystStageMutation);
      }

      // Operation 2: Update stage depending on stage type
      const activities = await updateStage(
        interval,
        stage,
        userName,
        status,
        activityMutation,
        analystStageMutation
      );

      // Operation 3: Open or Close the interval depending on its status
      if (activities.length === 0) return;
      openOrCloseStage(
        dispatch,
        interval,
        stage,
        status,
        isActivityInOpenStageInterval,
        openActivityNames,
        activities
      );
      const workflowPanelElement = workflowPanel[0] as HTMLElement;
      // On open or close focus is being lost, apply focus back to workflow panel
      workflowPanelElement?.focus();
    },
    [
      activityMutation,
      analystStageMutation,
      dispatch,
      openActivityNames,
      openIntervalName,
      status,
      userName,
      workflow?.stages,
      workflowPanel,
      workflowTimeRange
    ]
  );
}
