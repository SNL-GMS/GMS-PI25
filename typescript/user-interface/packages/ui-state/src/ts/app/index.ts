export * from './api';
export * from './history';
export * from './hooks';
export * from './query';
export * from './redux-provider';
export * from './sanitization';
export * from './state';
export * from './store';
export * from './subscription';
export * from './util';
