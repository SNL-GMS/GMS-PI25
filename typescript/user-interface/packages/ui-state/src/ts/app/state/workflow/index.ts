export * from './operations';
export * from './selectors';
export * from './types';
export * from './workflow-slice';
