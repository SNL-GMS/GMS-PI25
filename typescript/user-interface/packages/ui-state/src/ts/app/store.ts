import { IS_NODE_ENV_PRODUCTION, NODE_ENV } from '@gms/common-util';
import { getElectron, getElectronEnhancer, UILogger } from '@gms/ui-util';
import { configureStore } from '@reduxjs/toolkit';
import { createLogger } from 'redux-logger';
import type { Config } from 'redux-state-sync';
import {
  createStateSyncMiddleware,
  initStateWithPrevTab,
  withReduxStateSync
} from 'redux-state-sync';

import {
  eventManagerApiSlice,
  processingConfigurationApiSlice,
  processingStationApiSlice,
  signalEnhancementConfigurationApiSlice,
  sohAceiApiSlice,
  ssamControlApiSlice,
  stationDefinitionSlice,
  systemEventGatewayApiSlice,
  systemMessageDefinitionApiSlice,
  userManagerApiSlice,
  workflowApiSlice
} from './api';
import { eventConflictMiddleware } from './api/data/event/event-conflict-middleware';
import {
  eventRemovedMiddleware,
  eventUndoRedoMiddleware,
  historyAddAction,
  historyMiddleware,
  signalDetectionRemovedMiddleware,
  undoRedoMiddleware
} from './history';
import { reducer } from './reducer';
import { reduxDevToolsGlobalConfig } from './sanitization';
import { buildDenyList } from './sanitization/deny-list';
import { buildActionSanitizer, sanitizeState } from './sanitization/sanitizers';
import { getDevToolsConfigLogs } from './sanitization/util';
import { userSessionActions } from './state';
import {
  establishWsConnection,
  registerConnectionStatusCallback
} from './subscription/subscription';

const logger = UILogger.create('GMS_LOG_REDUX_STORE', process.env.GMS_LOG_REDUX_STORE || 'info');

const reduxStateSyncConfig: Config = {
  blacklist: [
    // to not sync the history add to prevent duplicates since the middleware is running in all windows
    historyAddAction.type
  ]
};

const buildStore = () => {
  const DISABLE_REDUX_STATE_SYNC = reduxDevToolsGlobalConfig.GMS_DISABLE_REDUX_STATE_SYNC.enabled;

  logger.info(
    `Configuring Redux store with the following properties: \n` +
      `NODE_ENV:${NODE_ENV}\n${getDevToolsConfigLogs()} `
  );

  const store = configureStore({
    reducer: !DISABLE_REDUX_STATE_SYNC ? withReduxStateSync(reducer) : reducer,
    devTools:
      !IS_NODE_ENV_PRODUCTION && !reduxDevToolsGlobalConfig.GMS_DISABLE_REDUX_DEV_TOOLS.enabled
        ? {
            trace: reduxDevToolsGlobalConfig.GMS_ENABLE_REDUX_TRACE.enabled,
            actionsDenylist: buildDenyList(),
            actionSanitizer: buildActionSanitizer(),
            stateSanitizer: sanitizeState
          }
        : false,
    middleware: getDefaultMiddleware => {
      const middlewares = getDefaultMiddleware({
        thunk: true,
        immutableCheck: reduxDevToolsGlobalConfig.GMS_ENABLE_REDUX_IMMUTABLE_CHECK.enabled,
        serializableCheck: reduxDevToolsGlobalConfig.GMS_ENABLE_REDUX_SERIALIZABLE_CHECK.enabled
      })
        // add the listener middleware to the store
        // NOTE: since this can receive actions with functions inside it should go before the serialize check middleware
        // ! the order of the middle matters is important!
        .prepend(undoRedoMiddleware.middleware)
        .prepend(eventUndoRedoMiddleware.middleware)
        .prepend(historyMiddleware.middleware)
        .prepend(eventRemovedMiddleware.middleware)
        .prepend(signalDetectionRemovedMiddleware.middleware)
        .prepend(eventConflictMiddleware.middleware)
        // ! custom middlewares must be prepended before this line
        .concat(systemEventGatewayApiSlice.middleware)
        .concat(eventManagerApiSlice.middleware)
        .concat(processingConfigurationApiSlice.middleware)
        .concat(processingStationApiSlice.middleware)
        .concat(sohAceiApiSlice.middleware)
        .concat(ssamControlApiSlice.middleware)
        .concat(signalEnhancementConfigurationApiSlice.middleware)
        .concat(stationDefinitionSlice.middleware)
        .concat(systemMessageDefinitionApiSlice.middleware)
        .concat(userManagerApiSlice.middleware)
        .concat(workflowApiSlice.middleware);

      if (!DISABLE_REDUX_STATE_SYNC) {
        middlewares.push(createStateSyncMiddleware(reduxStateSyncConfig));
      }

      // ! the logger should always be the last middleware added
      // enable the Redux logger only if `GMS_ENABLE_REDUX_LOGGER` is set to true
      if (reduxDevToolsGlobalConfig.GMS_ENABLE_REDUX_LOGGER.enabled) {
        const reduxLogger = createLogger({
          collapsed: true,
          duration: true,
          timestamp: false,
          level: 'info',
          logger: console,
          logErrors: true,
          diff: false
        });
        middlewares.push(reduxLogger);
      }
      return middlewares;
    },
    enhancers:
      getElectron() && getElectronEnhancer()
        ? [
            // must be placed after the enhancers which dispatch
            // their own actions such as redux-thunk or redux-saga
            getElectronEnhancer()({
              dispatchProxy: a => store.dispatch(a)
            })
          ]
        : []
  });

  // initialize state using any previous state from other tabs
  if (!DISABLE_REDUX_STATE_SYNC) {
    initStateWithPrevTab(store);
  }

  // store is created connect to System Event gateway and set callback to set connection status
  establishWsConnection();
  const updateConnectionStatus = (connected: boolean): void => {
    if (store) {
      store.dispatch(userSessionActions.setConnected(connected));
    }
  };
  registerConnectionStatusCallback(updateConnectionStatus, store.getState().app.userSession);

  return store;
};

export type ReduxStoreType = ReturnType<typeof buildStore>;

export interface GMSWindow extends Window {
  ReduxStore: ReduxStoreType;
}

const getGmsStore = (): ReturnType<typeof buildStore> => {
  const gmsWindow = (window as unknown) as GMSWindow;
  // ! ensure that only one Redux store instance is created
  if (gmsWindow.ReduxStore == null) {
    gmsWindow.ReduxStore = buildStore();
  }
  return gmsWindow.ReduxStore;
};

/**
 * Creates the Redux application store, which is stored on the window object in order to avoid any chance for
 * multiple stores to be created (if multiple copies of the module are included in the bundle, for example).
 *
 * @returns the Redux store for the application
 */
export const getStore = (): ReduxStoreType => getGmsStore();

/**
 * Returns the dispatch function of the Redux store/
 *
 * @returns the Redux dispatch function
 */
export const getDispatch = () => getGmsStore().dispatch;

/**
 * The application state.
 * Infer the `AppState` types from the store itself
 */
export type AppState = ReturnType<typeof reducer>;

/**
 * The keys of the application state.
 */
export type AppStateKeys = keyof AppState;

/**
 * The application dispatched (typed).
 * Infer the `AppDispatch` types from the store itself
 */
export type AppDispatch = ReturnType<typeof getDispatch>;
