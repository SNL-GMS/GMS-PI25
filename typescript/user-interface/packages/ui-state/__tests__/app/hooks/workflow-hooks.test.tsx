import { WorkflowTypes } from '@gms/common-model';
import { openIntervalName, user } from '@gms/common-model/__tests__/__data__';
import { renderHook } from '@testing-library/react-hooks';
import React from 'react';
import { Provider } from 'react-redux';

import {
  userSessionActions,
  useUpdateActivityIntervalStatusMutation,
  useUpdateStageIntervalStatusMutation
} from '../../../src/ts/app';
import { useSetInterval, useStageId } from '../../../src/ts/app/hooks/workflow-hooks';
import { workflowActions } from '../../../src/ts/app/state/workflow/workflow-slice';
import { getStore } from '../../../src/ts/app/store';
import { activityInterval, interactiveAnalysisStageInterval } from '../api/workflow/sample-data';

const store = getStore();

jest.mock('../../../src/ts/app/state/workflow/operations', () => {
  const actual = jest.requireActual('../../../src/ts/app/state/workflow/operations');

  return {
    ...actual,
    openOrCloseStage: jest.fn()
  };
});

jest.mock('../../../src/ts/app/api/workflow/workflow-api-slice', () => {
  const actual = jest.requireActual('../../../src/ts/app/api/workflow/workflow-api-slice');

  const mockActivityMutation = jest.fn();
  const mockAnalystStageMutation = jest.fn();

  return {
    ...actual,
    useUpdateActivityIntervalStatusMutation: () => [mockActivityMutation],
    useUpdateStageIntervalStatusMutation: () => [mockAnalystStageMutation],
    useWorkflowQuery: jest.fn(() => ({
      isSuccess: true,
      data: {
        stages: [
          {
            name: openIntervalName,
            mode: WorkflowTypes.StageMode.INTERACTIVE,
            activities: [
              {
                name: activityInterval.intervalId.definitionId.name,
                stationGroup: { name: 'mockStationGroup' }
              }
            ]
          },
          {
            name: 'AL2',
            mode: WorkflowTypes.StageMode.AUTOMATIC,
            activities: [
              {
                name: activityInterval.intervalId.definitionId.name,
                stationGroup: { name: 'mockStationGroup2' }
              }
            ]
          }
        ]
      }
    }))
  };
});

describe('Workflow Hooks', () => {
  describe('useStageId', () => {
    beforeEach(() => {
      store.dispatch(workflowActions.setOpenActivityNames(['AL1 Event Review']));
      store.dispatch(workflowActions.setOpenIntervalName('AL1'));
      store.dispatch(
        workflowActions.setTimeRange({ startTimeSecs: 1669150800, endTimeSecs: 1669154400 })
      );
    });
    it('returns the stage id if one is found', () => {
      const renderedStageIdResult = renderHook<{ children }, WorkflowTypes.IntervalId>(useStageId, {
        wrapper: ({ children }) => <Provider store={store}>{children}</Provider>
      });
      const renderedStageId = renderedStageIdResult.result.current;
      expect(renderedStageId).toMatchObject<WorkflowTypes.IntervalId>({
        startTime: 1669150800,
        definitionId: {
          name: 'AL1'
        }
      });
    });
    it('returns a newly created stage interval if no interval of the open type is found', () => {
      store.dispatch(workflowActions.setOpenIntervalName('OPENED_WITH_OPEN_ANYTHING')); // there is no OPENED_WITH_OPEN_ANYTHING interval in the mock data
      const renderedStageIdResult = renderHook<{ children }, WorkflowTypes.IntervalId>(useStageId, {
        wrapper: ({ children }) => <Provider store={store}>{children}</Provider>
      });
      const renderedStageId = renderedStageIdResult.result.current;
      expect(renderedStageId).toMatchObject<WorkflowTypes.IntervalId>({
        startTime: 1669150800,
        definitionId: {
          name: 'OPENED_WITH_OPEN_ANYTHING'
        }
      });
    });
    it('returns a newly created stage interval id if no interval with the expected timeRange is found', () => {
      store.dispatch(workflowActions.setOpenIntervalName('AL1'));
      store.dispatch(
        workflowActions.setTimeRange({ startTimeSecs: 1669150801, endTimeSecs: 1669154401 }) // time range does not match
      );
      const renderedStageIdResult = renderHook<{ children }, WorkflowTypes.IntervalId>(useStageId, {
        wrapper: ({ children }) => <Provider store={store}>{children}</Provider>
      });
      const renderedStageId = renderedStageIdResult.result.current;
      expect(renderedStageId).toMatchObject<WorkflowTypes.IntervalId>({
        startTime: 1669150801,
        definitionId: {
          name: 'AL1'
        }
      });
    });
    it('returns undefined if no open interval name is set', () => {
      store.dispatch(workflowActions.setOpenIntervalName(undefined));
      store.dispatch(
        workflowActions.setTimeRange({ startTimeSecs: 1669150800, endTimeSecs: 1669154400 })
      );
      const renderedStageIdResult = renderHook<{ children }, WorkflowTypes.IntervalId>(useStageId, {
        wrapper: ({ children }) => <Provider store={store}>{children}</Provider>
      });
      const renderedStageId = renderedStageIdResult.result.current;
      expect(renderedStageId).toBeUndefined();
    });
    it('returns undefined if no open time range is set', () => {
      store.dispatch(workflowActions.setOpenIntervalName('AL1'));
      store.dispatch(
        workflowActions.setTimeRange({
          startTimeSecs: null,
          endTimeSecs: null
        })
      );
      const renderedStageIdResult = renderHook<{ children }, WorkflowTypes.IntervalId>(useStageId, {
        wrapper: ({ children }) => <Provider store={store}>{children}</Provider>
      });
      const renderedStageId = renderedStageIdResult.result.current;
      expect(renderedStageId).toBeUndefined();
    });
  });

  describe('useSetInterval', () => {
    beforeEach(() => {
      store.dispatch(
        userSessionActions.setAuthenticationStatus({
          userName: user,
          authenticated: true,
          authenticationCheckComplete: true,
          failedToConnect: false
        })
      );
      store.dispatch(workflowActions.setOpenIntervalName(openIntervalName));
      store.dispatch(workflowActions.setOpenActivityNames(['AL1 Event Review']));
      store.dispatch(
        workflowActions.setTimeRange({ startTimeSecs: 1669150800, endTimeSecs: 1669154400 })
      );
      jest.clearAllMocks();
    });
    it('handles setting interval status with a stage interval', async () => {
      const [mockAnalystStageMutation] = useUpdateStageIntervalStatusMutation();
      const [mockActivityMutation] = useUpdateActivityIntervalStatusMutation();

      const status: WorkflowTypes.IntervalStatus = WorkflowTypes.IntervalStatus.IN_PROGRESS;
      const { result } = renderHook<
        { children },
        (interval: WorkflowTypes.ActivityInterval | WorkflowTypes.StageInterval) => Promise<void>
      >(() => useSetInterval(status), {
        wrapper: ({ children }) => <Provider store={store}>{children}</Provider>
      });

      expect(result.current).toBeDefined();
      expect(result.current).toMatchSnapshot();
      await result.current(interactiveAnalysisStageInterval);
      expect(mockAnalystStageMutation).toHaveBeenCalled();
      expect(mockActivityMutation).toHaveBeenCalledTimes(0);
    });

    it('handles setting interval status with an activity interval', async () => {
      const [mockAnalystStageMutation] = useUpdateStageIntervalStatusMutation();
      const [mockActivityMutation] = useUpdateActivityIntervalStatusMutation();

      const status: WorkflowTypes.IntervalStatus = WorkflowTypes.IntervalStatus.IN_PROGRESS;
      const { result } = renderHook<
        { children },
        (interval: WorkflowTypes.ActivityInterval | WorkflowTypes.StageInterval) => Promise<void>
      >(() => useSetInterval(status), {
        wrapper: ({ children }) => <Provider store={store}>{children}</Provider>
      });

      expect(result.current).toBeDefined();
      expect(result.current).toMatchSnapshot();
      await result.current(activityInterval);
      expect(mockActivityMutation).toHaveBeenCalled();
      expect(mockAnalystStageMutation).toHaveBeenCalledTimes(1);
    });
  });
});
