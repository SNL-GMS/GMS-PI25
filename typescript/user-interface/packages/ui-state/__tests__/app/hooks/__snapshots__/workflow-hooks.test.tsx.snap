// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Workflow Hooks useSetInterval handles setting interval status with a stage interval 1`] = `[Function]`;

exports[`Workflow Hooks useSetInterval handles setting interval status with an activity interval 1`] = `[Function]`;
